flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/datastream/DataStreamAggregate.scala

```diff
@@ -0,0 +1,261 @@
+    val prepareOpName = s"prepare select: ($aggString)"
+    val mappedInput = inputDS
+      .map(aggregateResult._1)
+      .name(prepareOpName)
+
+    val groupReduceFunction = aggregateResult._2
+    val rowTypeInfo = new RowTypeInfo(fieldTypes)
+
+    val result = {
+      // grouped / keyed aggregation
+      if (groupingKeys.length > 0) {
+        val aggOpName = s"groupBy: (${groupingToString(inputType, grouping)}), " +
+          s"window: ($window), " +
+          s"select: ($aggString)"
+        val aggregateFunction = new AggregateWindowFunction(groupReduceFunction)
+
+        val keyedStream = mappedInput.keyBy(groupingKeys: _*)
+
+        val windowedStream = createKeyedWindowedStream(window, keyedStream)
+          .asInstanceOf[WindowedStream[Row, Tuple, DataStreamWindow]]
+
+          windowedStream
+            .apply(aggregateFunction)
+            .returns(rowTypeInfo)
+            .name(aggOpName)
+            .asInstanceOf[DataStream[Any]]
+      }
+      // global / non-keyed aggregation
+      else {
+        val aggOpName = s"window: ($window), select: ($aggString)"
+        val aggregateFunction = new AggregateAllWindowFunction(groupReduceFunction)
+
+        val windowedStream = createNonKeyedWindowedStream(window, mappedInput)
+          .asInstanceOf[AllWindowedStream[Row, DataStreamWindow]]
+
+        windowedStream
+            .apply(aggregateFunction)
+            .returns(rowTypeInfo)
+            .name(aggOpName)
+            .asInstanceOf[DataStream[Any]]
+      }
+    }
+
+    // if the expected type is not a Row, inject a mapper to convert to the expected type
+    expectedType match {
+      case Some(typeInfo) if typeInfo.getTypeClass != classOf[Row] =>
+        val mapName = s"convert: (${getRowType.getFieldNames.asScala.toList.mkString(", ")})"
+        result.map(getConversionMapper(
+          config = config,
+          nullableInput = false,
+          inputType = rowTypeInfo.asInstanceOf[TypeInformation[Any]],
+          expectedType = expectedType.get,
+          conversionOperatorName = "DataStreamAggregateConversion",
+          fieldNames = getRowType.getFieldNames.asScala
+        ))
+        .name(mapName)
+      case _ => result
+    }
+  }
+
+}
+
+object DataStreamAggregate {
+
+  private def createKeyedWindowedStream(groupWindow: LogicalWindow, stream: KeyedStream[Row, Tuple])
+    : WindowedStream[Row, Tuple, _ <: DataStreamWindow] = groupWindow match {
+
+    case ProcessingTimeTumblingGroupWindow(_, size) if isTimeInterval(size.resultType) =>
+      stream.window(TumblingProcessingTimeWindows.of(asTime(size)))
+
+    case ProcessingTimeTumblingGroupWindow(_, size) =>
+      stream.countWindow(asCount(size))
+
+    case EventTimeTumblingGroupWindow(_, _, size) if isTimeInterval(size.resultType) =>
+      stream
+        .window(TumblingEventTimeWindows.of(asTime(size)))
+
+    case EventTimeTumblingGroupWindow(_, _, size) =>
+      stream.countWindow(asCount(size))
```

Comment:
I disabled the tests and throw exception at those locations. This should be fixed after this PR has been merged.